### PR TITLE
[DI][DX] Throw exception on some ContainerBuilder methods used from extensions

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no (unlikely, that would mean there already was an issue in userland code)
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #24282 
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

**Comment by iltar** [comment]
Yes, this is what I was referring to! This would reduce the amount of confusion and wtfs when working with the container.

File: src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php [hunk 1]
```diff
@@ -121,3 +124,35 @@ public function getEnvPlaceholders()
...
+    {
+        throw new LogicException(sprintf('You cannot add compiler passes to the container builder from an extension. Add it from the main container builder (from a bundle %s::build() method for instance).', Bundle::class));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function registerExtension(ExtensionInterface $extension)
+    {
+        throw new LogicException('You cannot register a extension from another extension.');
```
Comment (sstok) [inline src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php]: A**n** extension

File: src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php [hunk 2]
```diff
@@ -121,3 +124,35 @@ public function getEnvPlaceholders()
...
+ * @internal
+ */
+class MergeExtensionConfigurationContainerBuilder extends ContainerBuilder
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function addCompilerPass(CompilerPassInterface $pass, $type = PassConfig::TYPE_BEFORE_OPTIMIZATION/*, int $priority = 0*/)
+    {
+        throw new LogicException(sprintf('You cannot add compiler passes to the container builder from an extension. Add it from the main container builder (from a bundle %s::build() method for instance).', Bundle::class));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php]: You cannot add compiler passes to the container builder from an extension. Add it from your bundle's "build()" method instead.

**Review by nicolas-grekas — APPROVED** [review]
with "wording" comments

**Comment by ogizanagi** [comment]
Updated. But I was referring to the "main container builder" for those using the DI component standalone, without HttpKernel (so no bundle). Not sure how we could turn this better, but only mentioning `Bundle::build` is not pretty accurate to me :/

File: src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php [hunk 3]
```diff
@@ -121,3 +124,35 @@ public function getEnvPlaceholders()
...
+ * @internal
+ */
+class MergeExtensionConfigurationContainerBuilder extends ContainerBuilder
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function addCompilerPass(CompilerPassInterface $pass, $type = PassConfig::TYPE_BEFORE_OPTIMIZATION/*, int $priority = 0*/)
+    {
+        throw new LogicException('You cannot add compiler passes to the container builder from an extension. Add it from your bundle\'s "build()" method instead.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php]: so, would this be better?
`You cannot add compiler pass "%s" from extension "%s". Compiler passes must be registered before the container is compiled.`

File: src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php [hunk 4]
```diff
@@ -121,3 +124,35 @@ public function getEnvPlaceholders()
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function registerExtension(ExtensionInterface $extension)
+    {
+        throw new LogicException('You cannot register an extension from another extension.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php]: `You cannot register extension "%s" from "%s". Extensions must be registered before the container is compiled.`

File: src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php [hunk 5]
```diff
@@ -121,3 +124,35 @@ public function getEnvPlaceholders()
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function compile($resolveEnvPlaceholders = false)
+    {
+        throw new LogicException(sprintf('Cannot compile a %s instance.', __CLASS__));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php]: Cannot compile the container in extension "%s".
(no need to ref `__CLASS__`, that'd be confusing I think.

**Comment by ogizanagi** [comment]
Much better! Thanks @nicolas-grekas

**Pull request merged**